src/Development/IDE/Core/Compile.hs

```diff
@@ -148,11 +153,17 @@ typecheckModule (IdeDefer defer) hsc pm = do
     where
         demoteIfDefer = if defer then demoteTypeErrorsToWarnings else id
 
+modifyPLS :: DynLinker -> (PersistentLinkerState -> IO (PersistentLinkerState, a)) -> IO a
+modifyPLS dl f =
+  modifyMVar (dl_mpls dl) (fmapFst pure . f . fromMaybe undefined)
+  where fmapFst f = fmap (\(x, y) -> (f x, y))
+
 tcRnModule :: GhcMonad m => ParsedModule -> m TcModuleResult
 tcRnModule pmod = do
   let ms = pm_mod_summary pmod
   hsc_env <- getSession
   let hsc_env_tmp = hsc_env { hsc_dflags = ms_hspp_opts ms }
+  liftIO $ unload hsc_env_tmp []
```

Comment:
IIUC:

- In 8.8.x unload is buggy and can trigger the crash outlined in GHC #16525
- In 8.10.x unload does nothing and we rely on the system linker loading new objects with a newest-one-wins behaviour. But will it unload old ones? Probably not
- In 9.2 (or hopefully 9.0) unload will be fixed and reenabled and all will be good

Reply:
Yes, your understanding is largely correct.

> In 8.10.x unload does nothing

It also clears the `lookupSymbol` cache for the external interpreter, and also clears the `PersistentLinkerState` (mainly the `objs_loaded` field which is used by `getLinkDeps` to figure out which linkables (from the `HomePackageTable`) we need to load.